core/src/main/scala/cats/Eval.scala

```diff
@@ -134,23 +136,57 @@ final case class Now[A](value: A) extends Eval[A] {
  * by the closure) will not be retained, and will be available for
  * garbage collection.
  */
-final class Later[A](f: () => A) extends Eval[A] {
-  private[this] var thunk: () => A = f
-
-  // The idea here is that `f` may have captured very large
-  // structures, but produce a very small result. In this case, once
-  // we've calculated a value, we would prefer to be able to free
-  // everything else.
-  //
-  // (For situations where `f` is small, but the output will be very
-  // expensive to store, consider using `Always`.)
-  lazy val value: A = {
-    val result = thunk()
-    thunk = null // scalastyle:off
-    result
-  }
+final class Later[A](f: () => A) extends Eval[A] with Externalizable {
+
+  // we use lock to prevent possible problems with someone else
+  // synchronizing on 'this'.
+  private[this] val lock = new Object()
+  private[this] var thunk: Function0[A] = f
+  private[this] var result: A = _
+
+  final def value: A =
+    if (thunk == null) {
+      result
+    } else lock.synchronized {
+      if (thunk == null) result else {
```

Comment:
So I think the reason this isn't a problem is because we are only setting `thunk = null` after `result = thunk()` has fully initialized the `result: A` value. This means that we don't have to worry about someone else returning a partially-evaluated `result` value. I could be wrong though. I looked at this code with @johnynek -- maybe he can weigh in here too?